# Day 13: Care Package

As you ponder the solitude of space and the ever-increasing three-hour roundtrip for messages between you and Earth, you notice that the Space Mail Indicator Light is blinking. To help keep you sane, the Elves have sent you a care package.

It's a new game for the ship's arcade cabinet! Unfortunately, the arcade is **all the way** on the other end of the ship. Surely, it won't be hard to build your own - the care package even comes with schematics.

The arcade cabinet runs Intcode software like the game the Elves sent (your puzzle input). It has a primitive screen capable of drawing square **tiles** on a grid. The software draws tiles to the screen with output instructions: every three output instructions specify the `x` position (distance from the left), `y` position (distance from the top), and `tile id`. The `tile id` is interpreted as follows:
- `0` is an **empty** tile. No game object appears in this tile.
- `1` is a **wall** tile. Walls are indestructible barriers.
- `2` is a **block** tile. Blocks can be broken by the ball.
- `3` is a **horizontal paddle** tile. The paddle is indestructible.
- `4` is a **ball** tile. The ball moves diagonally and bounces off objects.

For example, a sequence of output values like `1,2,3,6,5,4` would draw a **horizontal paddle** tile (`1` tile from the left and `2` tiles from the top) and a **ball** tile (`6` tiles from the left and `5` tiles from the top).

Start the game. **How many block tiles are on the screen when the game exits?**

In [1]:
from pathlib import Path

input_path = Path("..") / "inputs" / "day_13.txt"
inputs = input_path.read_text()
inputs = inputs.split(",")
inputs = [int(input_) for input_ in inputs]

In [2]:
def make_chunks(lst, n):
    """Yield successive n-sized chunks from lst."""
    for i in range(0, len(lst), n):
        yield lst[i:i + n]

In [3]:
from collections import Counter

from computer import Computer

In [4]:
computer = Computer(inputs, [])
computer.run()
results = computer.outputs

chunks = make_chunks(results, 3)
tiles = [chunk[2] for chunk in chunks]
Counter(tiles)[2]

363

Answer to Part 1: **363**

## Part 2

The game didn't run because you didn't put in any quarters. Unfortunately, you did not bring any quarters. Memory address `0` represents the number of quarters that have been inserted; set it to `2` to play for free.

The arcade cabinet has a joystick that can move left and right. The software reads the position of the joystick with input instructions:
- If the joystick is in the **neutral position**, provide `0`.
- If the joystick is **tilted to the left**, provide `-1`.
- If the joystick is **tilted to the right**, provide `1`.

The arcade cabinet also has a segment display capable of showing a single number that represents the player's current score. When three output instructions specify `X=-1, Y=0`, the third output instruction is not a tile; the value instead specifies the new score to show in the segment display. For example, a sequence of output values like `-1,0,12345` would show `12345` as the player's current score.

Beat the game by breaking all the blocks. **What is your score after the last block is broken?**

In [5]:
def find_ball(computer):
    chunks = list(make_chunks(computer.outputs, 3))
    chunks.reverse()
    for chunk in chunks:
        if chunk[2] == 4:
            c = dict(zip(["x", "y"], chunk[:2]))
            return c


def find_paddle(computer):
    chunks = list(make_chunks(computer.outputs, 3))
    chunks.reverse()
    for chunk in chunks:
        if chunk[2] == 3:
            c = dict(zip(["x", "y"], chunk[:2]))
            return c

def find_score(computer):
    chunks = list(make_chunks(computer.outputs, 3))
    chunks.reverse()
    for chunk in chunks:
        if chunk[0] == -1 and chunk[1] == 0:
            return chunk[2]
        
def find_blocks(computer):
    chunks = list(make_chunks(computer.outputs, 3))
    chunks.reverse()
    for chunk in chunks:
        if chunk[2] == 2:
            yield chunk[0], chunk[1]

def find_clear(computer):
    chunks = list(make_chunks(computer.outputs, 3))
    chunks.reverse()
    for chunk in chunks:
        if chunk[2] == 0:
            yield chunk[0], chunk[1]

In [6]:
computer = Computer(inputs, [0])
computer.run()
blocks = find_blocks(computer)
blocks = set(blocks)

In [7]:
playing_inputs = inputs.copy()

In [8]:
playing_inputs[0] = 2

In [9]:
computer = Computer(playing_inputs, [0])

In [10]:
remaining = len(blocks)
while remaining:
    try:
        computer.run()
    except IndexError:
        pass
    ball = find_ball(computer)
    paddle = find_paddle(computer)
    score = find_score(computer)
    cleared = list(find_clear(computer))
    
    remaining = [i for i in blocks if i not in cleared]
    if ball["x"] > paddle["x"]:
        computer.inputs.append(1)
    elif ball["x"] < paddle["x"]:
        computer.inputs.append(-1)
    else:
        computer.inputs.append(0)
print(score)

17159


Answer to Part 2: **17159**